### ISSUE TITLE
parent first strategy is broken

### ISSUE THREAD

**Issue by asafbennatan** [pf4j/pf4j#381]
Hi,
having an issue with parent first strategy in PluginClassLoader,
seems like catch clause is catching ClassCastException instead of ClassNotFoundException:
```

else {
                // try to load from parent
                try {
                    return super.loadClass(className);
                } catch (ClassCastException e) {
                    // try next step
                }

```
also the issue here is that extensions.idx is being loaded from parent causing wrong extensions.idx to be loaded , i suggest solving this by adding reserved resource names META-INF/extensions.idx and META-INF/MANIFEST.MF ( in getResource) , these will be loaded locally even if strategy is parent first

**Comment by decebals** [comment]
> seems like catch clause is catching ClassCastException instead of ClassNotFoundException:

Your observation is good. Can you create a PR? If not, no problem, I'll fix it.

> also the issue here is that extensions.idx is being loaded from parent causing wrong extensions.idx to be loaded , i suggest solving this by adding reserved resource names META-INF/extensions.idx and META-INF/MANIFEST.MF ( in getResource) , these will be loaded locally even if strategy is parent first

This problem is debatable. In general the parent class loader for a plugin is an `AppClassLoader` (or something similar) and you don't have an `extensions.idx` in parent. Can you give me a concrete use case in which you encounter the case exposed by you in this issue?

**Comment by asafbennatan** [comment]
1.ill try creating a PR 
2. what about extensions which are present in base project ( i remember that in the docs it says somewhere that you may create base project extensions and create plugins out of them later on )? extensions.idx should be in base classloader to support it or not ? if not then you are right , base should not have extensions.idx

**Comment by asafbennatan** [comment]
pull request made: https://github.com/pf4j/pf4j/pull/383 
note that it does not pass the build due to demo failing the travis build ( not sure why , i just took the one on master) , 
i did not see an contribution guide so hopefully this is ok by the project standards

**Comment by decebals** [comment]
> 1.ill try creating a PR

Thanks!

> 2. what about extensions which are present in base project ( i remember that in the docs it says somewhere that you may create base project extensions and create plugins out of them later on )? extensions.idx should be in base classloader to support it or not ? if not then you are right , base should not have extensions.idx

I need to check. Maybe it's a good idea to add a unit test for regression.

**Comment by decebals** [comment]
Solved by #384.
